### PR TITLE
8262002: java/lang/instrument/VerifyLocalVariableTableOnRetransformTest.sh failed with "TestCaseScaffoldException: DummyClassWithLVT did not match .class file"

### PR DESCRIPTION
**Pull request by alexmenkov**
The test actually failed starting from jdk13, but the error is masked by JDK-8264667 (and shell part of the test didn't verify result of the java execution)
The fix:
- updates JvmtiClassFileReconstituter to add attributes in the same order as javac does
- makes the test java instead of shell
- added workaround for JDK-8264667
- test code is ready to ignore order of attributes

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8262002](https://bugs.openjdk.java.net/browse/JDK-8262002): java/lang/instrument/VerifyLocalVariableTableOnRetransformTest.sh failed with "TestCaseScaffoldException: DummyClassWithLVT did not match .class file"

### Reviewers
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**)
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3426/head:pull/3426` \
`$ git checkout pull/3426`

Update a local copy of the PR: \
`$ git checkout pull/3426` \
`$ git pull https://git.openjdk.java.net/jdk pull/3426/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3426`

View PR using the GUI difftool: \
`$ git pr show -t 3426`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3426.diff">https://git.openjdk.java.net/jdk/pull/3426.diff</a>

</details>

### CODE REVIEW

**Comment by alexmenkov** [comment]
label remove core-libs

File: test/jdk/java/lang/instrument/ATransformerManagementTestCase.java [hunk 1]
```diff
@@ -117,9 +117,13 @@ public ATransformerManagementTestCase(String name)
...
             fTransformers.add(transformer);
         }
+        // workaroud for JDK-8264667: create log message before addTransformer()
+        String msg = "Added transformer " + transformer
+            + " with canRetransform=" + canRetransform;
         manager.addTransformer(transformer, canRetransform);
-        verbosePrint("Added transformer " + transformer
-            + " with canRetransform=" + canRetransform);
+//        verbosePrint("Added transformer " + transformer
+//            + " with canRetransform=" + canRetransform);
```
Comment (sspitsyn) [inline test/jdk/java/lang/instrument/ATransformerManagementTestCase.java:125]: I'd suggest to remove the commented out lines.

File: test/jdk/java/lang/instrument/VerifyLocalVariableTableOnRetransformTest.java [hunk 2]
```diff
@@ -120,8 +137,92 @@ public VerifyLocalVariableTableOnRetransformTest(String name)
...
+    private File originalTargetClassFile() {
+        String resourceName = fTargetClassName + ".class";
+        return new File(System.getProperty("test.classes", "."), resourceName);
+    }
+
+    private File transformClassFile() {
+        // This file will get created in the test execution
+        // directory so there is no conflict with the file
+        // in the test classes directory.
+        String resourceName = fTargetClassName + ".class";
```
Comment (sspitsyn) [inline test/jdk/java/lang/instrument/VerifyLocalVariableTableOnRetransformTest.java:152]: This name can be defined out of methods `originalTargetClassFile` and `transformClassFile`. The method name `transformClassFile` is confusing. I'd suggest to rename it to `transformedClassFile` or `modifiedClassFile`. Also, the name `originalTargetClassFile` can be shortened to `originalClassFile`.

File: test/jdk/java/lang/instrument/VerifyLocalVariableTableOnRetransformTest.java [hunk 3]
```diff
@@ -120,8 +137,92 @@ public VerifyLocalVariableTableOnRetransformTest(String name)
...
+        System.out.println(fTargetClassName + " did not match .class file");
+        System.out.println("Disassembly difference (" + f1 + " vs " + f2 +"):");
+        // compare 'javap -v' output for the class files
+        List<String> out1 = disassembleClassFile(f1);
+        List<String> out2 = disassembleClassFile(f2);
+        boolean different = false;
+        boolean orderChanged = false;
+        int lineNum = 0;
+        for (String line: out1) {
+            if (expectedDiffenent(line)) {
```
Comment (sspitsyn) [inline test/jdk/java/lang/instrument/VerifyLocalVariableTableOnRetransformTest.java:186]: Is it possible to use the `expectedDiffenent()` as a filter to get only needed lines from `disassembleClassFile()`?

File: test/jdk/java/lang/instrument/ATransformerManagementTestCase.java (hunk 1 above)
Comment (alexmenkov) [reply test/jdk/java/lang/instrument/ATransformerManagementTestCase.java:125]: fixed

File: test/jdk/java/lang/instrument/VerifyLocalVariableTableOnRetransformTest.java (hunk 2 above)
Comment (alexmenkov) [reply test/jdk/java/lang/instrument/VerifyLocalVariableTableOnRetransformTest.java:152]: done

File: test/jdk/java/lang/instrument/VerifyLocalVariableTableOnRetransformTest.java (hunk 3 above)
Comment (alexmenkov) [reply test/jdk/java/lang/instrument/VerifyLocalVariableTableOnRetransformTest.java:186]: It's implemented this way to keep lineNum correct. But it was error here. Fixed now

Comment (alexmenkov) [reply test/jdk/java/lang/instrument/VerifyLocalVariableTableOnRetransformTest.java:186]: And fixed typo in the method name

**Comment by alexmenkov** [comment]
Ping

**Review by coleenp — APPROVED** [review]
Looks good and great to replace the shell file in the process.

**Review by sspitsyn — APPROVED** [review]
Hi Alex,
Looks good to me.
Thanks,
Serguei

**Pull request closed without merging**